Cluster management/newdashboard task

MaddyDev opened the pull request:
Added 'Cluster Status' task to the SQL Server big data cluster dashboard. 
This would copy the cluster-status notebook to users local machine and opens it up.
The notebook asks for the following inputs before executing the mssqlctl commands for login:
- cluster name
- controller username
- controller password
- controller endpoint, only when the cluster info isn't found in the kube config.

Also updated the bdc-endpoints widget to have all the endpoints as links as per Mihaela's request. And updated the service names following JRJ's email thread, also added a copy icon that copies the endpoint to the clipboard on his request. 

Below is what the dashboard looks like:

![image](https://user-images.githubusercontent.com/12754347/60131857-2de98500-974f-11e9-9542-744b1a2ab43f.png)

I'll update the cluster-status notebook once I get the final version from Alan Yu.

kevcunnane commented:
Did you get the copy icon from Smitha? It doesn't look super clear to me that it's a copy button...

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
+		endpointsArray.forEach(endpointInfo => {
+			endpointInfo.isHyperlink = true;
+			endpointInfo.hyperlink = 'https://' + endpointInfo.ipAddress + ':' + endpointInfo.port;
+
```
Nitpick: remove unneeded whitespace line

MaddyDev replied:
I did the format document since I got some hygiene errors when committing and I guess the spaces are added due to it. I'll see if I can remove them and commit

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
+		});
 		if (endpointsArray.length > 0) {
-			const managementProxyEp = endpointsArray.find(e => e.serviceName === 'management-proxy');
+			const managementProxyEp = endpointsArray.find(e => e.serviceName === 'management-proxy' || e.serviceName === 'mgmtproxy');
```
Did this change in a recent CTP?

MaddyDev replied:
Yes, I sent out an email regarding this and added you to it.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
 			if (managementProxyEp) {
-				endpointsArray.push(getCustomEndpoint(managementProxyEp, 'Grafana Dashboard', '/grafana'));
-				endpointsArray.push(getCustomEndpoint(managementProxyEp, 'Kibana Dashboard', '/kibana'));
+				endpointsArray.push(getCustomEndpoint(managementProxyEp, 'Metrics Dashboard', '/grafana'));
```
Please localize these endpoints since we control localization for them. For others that we get back directly from the APIs, we can't localize but these we are able to.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
 			if (gatewayEp) {
-				endpointsArray.push(getCustomEndpoint(gatewayEp, 'Spark History', '/gateway/default/sparkhistory'));
-				endpointsArray.push(getCustomEndpoint(gatewayEp, 'Yarn History', '/gateway/default/yarn'));
+				endpointsArray.push(getCustomEndpoint(gatewayEp, 'Spark Job Monitoring', '/gateway/default/sparkhistory'));
```
Both these endpoint names are also localizable.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
 				else {
 					const endpointCell = view.modelBuilder.text().withProperties<azdata.TextComponentProperties>({ value: endpointInfo.ipAddress + ':' + endpointInfo.port }).component();
-					endPointRow.addItem(endpointCell, { CSSStyles: { 'width': '70%' } });
+					endPointRow.addItem(endpointCell, { CSSStyles: { 'width': '60%', 'user-select': 'text' } });
```
Could we just make the entire table `user-select`able?

MaddyDev replied:
added the user-select to the container.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,47 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
-					endPointRow.addItem(endpointCell, { CSSStyles: { 'width': '70%' } });
+					endPointRow.addItem(endpointCell, { CSSStyles: { 'width': '60%', 'user-select': 'text' } });
 				}
+				const copyValueCell = view.modelBuilder.button().component();
```
Can we add hovertext for this? For accessibility, we need to make sure this clearly is marked as `Copy` and it helps all users.

MaddyDev replied:
![image](https://user-images.githubusercontent.com/12754347/59884600-b641e180-936d-11e9-9450-904da6a7ef3e.png)

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -187,6 +199,30 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
+		let friendlyName: string = name;
+		switch (name) {
+			case 'app-proxy':
+				friendlyName = 'Application Proxy';
```
Anything we're giving names for should be localized. Let's use the localize function here.

MaddyDev replied:
Added the localization

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,122 @@
+{
```
nitpick: let's use `tsg` for directory name if possible to minimize casing risks on different platforms.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -297,6 +336,21 @@ async function handleOpenNotebookTask(profile: azdata.IConnectionProfile): Promi
 
+async function handleOpenClusterStatusNotebookTask(profile: azdata.IConnectionProfile): Promise<void> {
+	const notebookRelativePath = 'mssql/notebooks/TSG/cluster-status.ipynb';
+	const notebookFullPath = path.join(__dirname, '../../', notebookRelativePath);
```
Does this logic work when the extension code is webpacked during full build? I am concerned relative lookup will fail. I believe going relative to the extension's root directory is the recommended way forward here. That's what we do for other resources in the notebook extension as far as I remember

MaddyDev replied:
I updated it to get the path relative to the extension's root.

kevcunnane commented on extensions/mssql/src/utils.ts:
```diff
@@ -29,6 +30,33 @@ export function getAppDataPath() {
+ */
+export function getTargetFileName(filePath: string, fileExtension: string): string {
+	const targetDirectory = os.homedir();
+	const baseName = path.basename(filePath, fileExtension);
```
Isn't this logic going to produce `cluster-status.ipynb-1.ipynb`? I just ran into the same issue, and used the below logic that doesn't require passing in the fileExtension, and would output `cluster-status-1.ipynb`. I recommend using the same here. Ideally we'd share the code but as this is in separate extensions, I guess we'll have to dupe for now.
https://github.com/microsoft/azuredatastudio/pull/6093/files

```ts
	private getUntitledUri(originalTitle: string): vscode.Uri {
		let title = originalTitle;
		let nextVal = 0;
		let ext = path.extname(title);
		while (!isEditorTitleFree(title)) {
			if (ext) {
				// Need it to be `Readme-0.txt` not `Readme.txt-0`
				let titleStart = originalTitle.slice(0, originalTitle.length - ext.length);
				title = `${titleStart}-${nextVal}${ext}`;
			} else {
				title = `${originalTitle}-${nextVal}`;
			}
			nextVal++;
		}
		return vscode.Uri.parse(`untitled:${title}`);
	}

... in a utils module:
export function isEditorTitleFree(title: string): boolean {
	let hasTextDoc = vscode.workspace.textDocuments.findIndex(doc => doc.isUntitled && doc.fileName === title) > -1;
	let hasNotebookDoc = azdata.nb.notebookDocuments.findIndex(doc => doc.isUntitled && doc.fileName === title) > -1;
	return !hasTextDoc && !hasNotebookDoc;
} 

```

MaddyDev replied:
path.basename only get's the filename without the extension. So it only get's cluster-status as the basename and extension '.ipynb' is added when forming the targetFileName. I tested this and it works. Do you still want me to change it and use your implementation for uniformity?

kevcunnane replied:
I'd prefer if you can skip passing in '.ipynb' as a hardcoded thing. Looking at path.basename, it's working as you pass in the extension and it knows to skip that part. Could you try:
```ts
	let fileExtension = path.extname(filePath); 
       const baseName = path.basename(filePath, fileExtension);
```
So that we skip this? In the future, it'd be nice that this can work for any file.

alanrenmsft commented:
IMPORTANT: please do a clear results on the notebook, i noticed it has output in it now.

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,167 @@
+            "cell_type": "code",
+            "source": "import sys, platform\r\n\r\nif platform.system()==\"Windows\":\r\n    user = ' --user'\r\nelse:\r\n    user = ''\r\n    \r\ncmd = f'{sys.executable} -m pip uninstall --yes mssqlctl-cli-storage'\r\ncmdOutput = !{cmd}\r\ncmd = f'{sys.executable} -m pip uninstall -r http://helsinki/browse/packages/python/aris-p-release-candidate-gb/mssqlctl/requirements.txt --yes'\r\n\r\ncmdOu ...
+            "metadata": {},
+            "outputs": [
```
Outputs still need to be cleared

MaddyDev replied:
I'll make sure to clear the output on the final copy that Alan Yu sends me with documentation.

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -137,35 +137,50 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
+				copyValueCell.onDidClick(() => {
+					vscode.env.clipboard.writeText(endpointInfo.hyperlink);
+				});
+				copyValueCell.title = 'Copy';
```
Localize this

MaddyDev replied:
I keep forgetting this, smh. On it. thanks again

kevcunnane commented on extensions/mssql/src/main.ts:
```diff
@@ -297,6 +339,21 @@ async function handleOpenNotebookTask(profile: azdata.IConnectionProfile): Promi
+	const notebookRelativePath = 'notebooks/tsg/cluster-status.ipynb';
+	const notebookFullPath = path.join(appContext.extensionContext.extensionPath, notebookRelativePath);
+	if (!Utils.fileExists(notebookFullPath)) {
+		vscode.window.showErrorMessage(localize('fileNotFound', 'Unable to find the file specified.'));
```
Minor: we're meant to use `""` for localized strings, e.g.
```suggestion
		vscode.window.showErrorMessage(localize('fileNotFound', "Unable to find the file specified"));
```
Also usually no need for `.` at the end in these messages.

MaddyDev replied:
I just followed the other localize examples in the file. I'll update them, Thanks!

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -22,146 +22,138 @@
-            ],
-            "execution_count": 15
+            "outputs": [],
+            "execution_count": 1
```
Please remove / reset execution_count to 0 for all cells 😄

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -22,146 +22,138 @@
-            ],
-            "execution_count": 11
+            "outputs": [],
+            "execution_count": 2
```
Looks like it's a bug that we don't clear this on clearing outputs, file a bug?

MaddyDev commented:
I have cleaned up everything- removed all the IP's and username/password from this notebook.
The only thing I overlooked is the url which is pointing to the internal one, just updated it to the correct one from https://docs.microsoft.com/en-us/sql/big-data-cluster/deploy-install-mssqlctl?view=sqlallproducts-allversions

Yes, I agree we need to review the notebooks with extra care since the content is JSON. But as Chris suggested, we should view it as a file and review it. I can schedule sometime but isn't this supposed to be available by tomorrow in the insiders?

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+            "cell_type": "markdown",
+            "source": "## **Dependencies**\r\n\r\n> This Notebook will try to install these dependencies for you.\r\n\r\n----------------------------------------------------------------\r\n<table>\r\n<colgroup>\r\n<col style=\"width: 10%\" />\r\n<col style=\"width: 10%\" />\r\n<col style=\"width: 10%\" />\r\n<col style=\"width: 10%\" />\r\n</colgroup>\r\n<thead>\r\n<tr class=\"header\">\r\n<th>To ...
```
Minor: why aren't we using markdown table? HTML isn't rendering in Github?

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+            "cell_type": "code",
+            "source": "#install pandas\r\ncmd = f'{sys.executable} -m pip show pandas'\r\ncmdOutput = !{cmd}\r\nif len(cmdOutput) > 0 and '0.24' in cmdOutput[1]:\r\n    print('Pandas required version is already installed!')\r\nelse:\r\n    pandasVersion = 'pandas==0.24.2'\r\n    cmd = f'{sys.executable} -m pip install {pandasVersion}'\r\n    cmdOutput = !{cmd}\r\n    print(f'\\nSuccess: Upgraded  ...
```
This is... unusual. Why fix on v0.24 of pandas? Once it gets bumped to 0.25 or higher this'll break. In an update can we get this logic fixed, and ideally just support a command to open Manage Packages with appropriate version?

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+            "cell_type": "markdown",
+            "source": "## **Log in to your big data cluster**\r\nTo view cluster status, you will need to connect to your big data cluster through mssqlctl. \r\n\r\nWhen you run this code cell, you will be prompted for:\r\n- Cluster name\r\n- Controller username\r\n- Controller password\r\n\r\nTo proceed:\r\n- **Click** on the input box\r\n- **Type** the login info\r\n- **Press** enter.\r\n\r\nIf ...
```
Can we just have format be `ipAddress:port`? The `00` format is confusing here, especially since standard is 3 digits not 2 for most IP parts.

kevcunnane commented on extensions/mssql/notebooks/TSG/cluster-status.ipynb:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+            "cell_type": "markdown",
+            "source": "## **Cluster Status**\r\nFor each cluster component below, running each code cell will generate a table. This table will include:\r\n\r\n----------------------------------------------------------------\r\n<table>\r\n<colgroup>\r\n<col style=\"width: 10%\" />\r\n<col style=\"width: 10%\" />\r\n</colgroup>\r\n<thead>\r\n<tr class=\"header\">\r\n<th>Column Name</th>\r\n<th>Des ...
```
All tables should just use markdown format. They don't show in other viewers (e.g. github) in the current form
